File: src/main/java/com/github/huoguangjin/multihighlight/highlight/TextAttributesFactory.java
package com.github.huoguangjin.multihighlight.highlight;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import com.github.huoguangjin.multihighlight.config.MultiHighlightConfig;
import com.github.huoguangjin.multihighlight.config.NamedTextAttr;

/**
 * Created by Rammer on 06/02/2017.
 */
public class TextAttributesFactory {

    private static final TextAttributesFactory INSTANCE = new TextAttributesFactory();

    private static int INDEX;

    public static TextAttributesFactory getInstance() {
        return INSTANCE;
    }

    private List<NamedTextAttr> namedTextAttrs;

    private TextAttributesFactory() {
        update();
    }

    public void update() {
        namedTextAttrs = MultiHighlightConfig.getInstance().getNamedTextAttrs();
    }

    @NotNull
    public NamedTextAttr get() {
        final List<NamedTextAttr> colors = this.namedTextAttrs;
        if (colors.isEmpty()) {
            return NamedTextAttr.IDE_DEFAULT; // NOTICE: remember me
        } else {
            return colors.get(INDEX++ % colors.size());
        }
    }
}

File: src/main/java/com/github/huoguangjin/multihighlight/config/TextAttributesFactory.kt
package com.github.huoguangjin.multihighlight.config

object TextAttributesFactory {

  private val namedTextAttrs: MutableList<NamedTextAttr> by lazy {
    MultiHighlightConfig.getInstance().namedTextAttrs.ifEmpty {
      listOf(NamedTextAttr.IDE_DEFAULT)
    }.toMutableList()
  }

  private var index = 0

  internal fun update(textAttrs: List<NamedTextAttr>) {
    namedTextAttrs.clear()
    namedTextAttrs.addAll(textAttrs.ifEmpty {
      listOf(NamedTextAttr.IDE_DEFAULT)
    })

    index %= namedTextAttrs.size
  }

  fun getTextAttrs(): List<NamedTextAttr> = namedTextAttrs

  fun getNextTextAttrIndex(): Int = index

  fun advanceTextAttrIndex(): Int {
    index = (index + 1) % namedTextAttrs.size
    return index
  }

  fun getNextTextAttr(): NamedTextAttr {
    return namedTextAttrs[index].also {
      advanceTextAttrIndex()
    }
  }
}
